dev-docs/RFCs/v7.x/imperative-api-rfc.md

```diff
@@ -0,0 +1,217 @@
+
+
+## API Design Constraints
+
+### ONE API Principle
+
+deck.gl is designed based on a "ONE API" principle, which is intended as a strong encouragement to API designers to ensure that we do not create irreconcilable forks in our API, as we continue to develop the React, Pure JS, scripting, JSON etc versions of the deck.gl API. Ideally a program should be portable between versions with very small differences, and learning how to use deck.gl in one version should translate directly to all versions.
+
+Therefore any proposals here need to consider this principle and justify the design in case of deviations. Of course, by taking the conversation one step up, and talking about imperative vs. reactive/functional paradigms we can argue that this support does not cater to one specific API version but rather to ensure that we support multiple programming paradigms. Still, these changes will make programs written in the various deck.gl API incarnations more different, so it is good to exercise constraint when possible.
+
+### Layers are Immutable
+
+The user would surely like to call `layer.setProps({...})` to partially update a layer but this is not possible in our current API model where layers are immutable descriptors. Instead of exploring the complexities of changing that model, a separate mechanism for updating layers is proposed.
+
+
+## Proposals
+
+This RFC proposes an `updateTriggers` overload (to set alternate semanatics) and new `Deck` method(s) (to partially update layers).
+
+
+### Layer.updateTriggers: true
+
+By setting `updateTriggers` to `true` (normally an object), an imperative application can force accessors to be value compared.
+
+This could be the default behavior when calling `Deck.updateLayers`.
+
+
+### Deck.updateLayers(layerMap : Object)
+
+Updates specified props in existing layers, without changing any other props. Effectively clones the existing layer, overwriting the specified props.
+
+```
+deck.updateLayers({
+  'scatterplot': {
+    ...newProps
+  }
+});
+```
+
+The `updateLayers` function takes a map with layer id keys, where each value is a map of props to update for that layer.
+
+* UpdateLayers can't remove existing layers
+
+Remarks:
+
+* If no corresponding layer exists, that update key will be ignored (a console warning may be emitted).
+* Updating child layers is currently not supported - this will create state inconsistencies and should happen through a common prop forwarding scheme for sublayers (so that the update goes through the top-level layer).
+* The applicatin's choice of whether to re-generate all props and layers or just update selected props in a few layers is mainly stylistic (deck.gl's highly optimized layer update engine typically makes the performance difference irrelevant).
+
+
+### Deck.addLayers(layers)
+
+> deck.updateLayers can't actually add layers...
+
+* updateLayers could interpret layer classes as new layers `id: new ScatterplotLayer(id, ...)`
```

Comment:
> Could you elaborate on interpret layer classes?

Reworded to clarify that supplying a layer class instance instead of object would cause the new layer to be added.